### ISSUE TITLE
Undefined variable with experimental patterns feature

### ISSUE THREAD

**Issue by DFelten** [Dart-Code/Dart-Code#4369]
**Describe the bug**

It seems that the new Dart 3 feature patterns are not working correctly with VSCode. The variable is not defined (see example). The code can be executed normally, only the analyzer shows an error. After removing the patterns flag, it's working. 

**To Reproduce**

1. Enable patterns via analysis_options.yaml

```yaml
analyzer:
  enable-experiment:
    - records
    - patterns
```

2. Create a dart method with the following code

```dart
final test = jsonDecode('{"test": "test"}') as Map<String, dynamic>;

print(test);
```

**Expected behavior**

Variable test should be defined for the analyser.

**Screenshots**

<img width="1165" alt="CleanShot 2023-01-31 at 07 02 39@2x" src="https://user-images.githubusercontent.com/1169185/215679127-8d02f69f-0751-4b1c-9dcb-03cda138fc3d.png">

**Please complete the following information:**
 - Operating System and version: MacOS Ventura 13.1
 - VS Code version:

1.75.0-insider
0615c4f98b07b51d3be6a357e22afec119490c5a
arm64

 - Dart extension version: v3.59.20230125 Pre-Release
 - Dart/Flutter SDK version: 2.19 / 3.7.0

**Comment by DanTup** [comment]
I don't seem to be able to reproduce this:

<img width="1792" alt="Screenshot 2023-01-31 at 10 51 47" src="https://user-images.githubusercontent.com/1078012/215740754-be1bb3b7-b1f4-428e-a8d9-d37b079fb5da.png">

Can you reproduce it in a newly created project that you could zip up and share so I have exactly the same files/pubspec/etc.?

I'm also unsure what the state of patterns is expected to be in stable 2.19 since it's still being actively worked on (and there have been many fixes after that version branched). @bwilkerson may have a better idea of this though.

**Comment by DFelten** [comment]
We also do not use it in production. I just wanted to have a look at the function. In two separate existing projects, we have the problem (Unfortunately they are private). However, in the test project it works.

Here is an excerpt from the code. I will try out here further, what is the reason that it is not possible here. 

![CleanShot 2023-01-31 at 12 35 11](https://user-images.githubusercontent.com/1169185/215749391-79a11845-eeab-4ea7-b637-d1ba5e37545b.png)

**Comment by DanTup** [comment]
Oh, I see - I can reproduce using your code above (that is, where the code is inside a `case`):

![Screenshot 2023-01-31 at 12 16 57](https://user-images.githubusercontent.com/1078012/215757450-9389663e-0aa6-45e2-a423-783db30c7bcb.png)

However, if I switch to the latest bleeding-edge SDK, the issue goes away:

![Screenshot 2023-01-31 at 12 17 29](https://user-images.githubusercontent.com/1078012/215757563-133dfc79-e29c-4481-978f-a0c0747b446b.png)

So I think this is a bug that has been fixed since 2.19 branched. I don't think 2.19 was intended to have complete support (at least from the analysis server) for patterns, and if you want to try this out seriously you'd have to use a a dev/bleeding edge build (which of course, I wouldn't recommend for production work 🙂).

**Comment by DanTup** [comment]
@bwilkerson thanks for confirming!

@DFelten I'll close this as I don't think there's anything to do - the issue you hit is already fixed in latest code.

If you do want to try patterns/etc. out early, you should use a pre-release build (such as [Dart 3 alpha](https://medium.com/dartlang/dart-3-alpha-f1458fb9d232) or [bleeding edge ("main") build](https://dart.dev/get-dart/archive#:~:text=x64%2Drelease.zip-,Main%20channel%20URL%20scheme,-The%20latest%20main)). Even then, you might/will still find things incomplete.
